From: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Date: Thu, 2 May 2024 13:38:44 +0200
Subject: [PATCH] media: cec: core: avoid confusing "transmit timed out" message
Message-ID: <70b13fcf5d43b567595fd7d4b25157ae793ab09c.1714649924.git.hverkuil-cisco@xs4all.nl>
Patch-mainline: Submitted, linux-media ML
References: CVE-2024-23848 bsc#1219104

If, when waiting for a transmit to finish, the wait is interrupted,
then you might get a "transmit timed out" message, even though the
transmit was interrupted and did not actually time out.

Set transmit_in_progress_aborted to true if the
wait_for_completion_killable() call was interrupted and ensure
that the transmit is properly marked as ABORTED.

Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Reported-by: Yang, Chenyuan <cy54@illinois.edu>
Fixes: 590a8e564c6ef ("media: cec: abort if the current transmit was canceled")
Closes: https://lore.kernel.org/linux-media/PH7PR11MB57688E64ADE4FE82E658D86DA09EA@PH7PR11MB5768.namprd11.prod.outlook.com/
Acked-by: Takashi Iwai <tiwai@suse.de>

---
 drivers/media/cec/cec-adap.c |    6 ++++++
 1 file changed, 6 insertions(+)

--- a/drivers/media/cec/cec-adap.c
+++ b/drivers/media/cec/cec-adap.c
@@ -392,6 +392,12 @@ int cec_thread_func(void *_adap)
 			goto unlock;
 		}
 
+		if (adap->transmit_in_progress_aborted) {
+			if (adap->transmitting)
+				cec_data_cancel(adap->transmitting);
+			adap->transmit_in_progress_aborted = false;
+			goto unlock;
+		}
 		if (adap->transmitting && timeout) {
 			/*
 			 * If we timeout, then log that. This really shouldn't
